src/MOI_wrapper/MOI_nonlinear.jl

```diff
@@ -0,0 +1,349 @@
+        :* => GRB_OPCODE_MULTIPLY,
+        :/ => GRB_OPCODE_DIVIDE,
+        # GRB_OPCODE_UMINUS        6
+        # GRB_OPCODE_SQUARE        7
+        :sqrt => GRB_OPCODE_SQRT,
+        :sin => GRB_OPCODE_SIN,
+        :cos => GRB_OPCODE_COS,
+        :tan => GRB_OPCODE_TAN,
+        :^ => GRB_OPCODE_POW,
+        :exp => GRB_OPCODE_EXP,
+        :log => GRB_OPCODE_LOG,
+        :log2 => GRB_OPCODE_LOG2,
+        :log10 => GRB_OPCODE_LOG10,
+        :logistic => GRB_OPCODE_LOGISTIC,
+    )
+else
+    const _OPCODE_MAP = Dict()
+end
+
+function MOI.supports_constraint(
+    ::Optimizer,
+    ::Type{MOI.ScalarNonlinearFunction},
+    ::Type{
+        <:Union{
+            MOI.LessThan{Float64},
+            MOI.GreaterThan{Float64},
+            MOI.EqualTo{Float64},
+        },
+    },
+)
+    return _supports_nonlinear()
+end
+
+function _info(
+    model::Optimizer,
+    c::MOI.ConstraintIndex{MOI.ScalarNonlinearFunction,S},
+) where {S}
+    if haskey(model.nl_constraint_info, c.value)
+        return model.nl_constraint_info[c.value]
+    end
+    return throw(MOI.InvalidIndex(c))
+end
+
+function _add_expression_tree_constant(
+    opcode::Vector{Cint},
+    data::Vector{Cdouble},
+    parent::Vector{Cint},
+    coeff::Float64,
+    index::Cint,
+)
+    append!(opcode, GRB_OPCODE_CONSTANT)
+    append!(data, coeff)
+    return append!(parent, index)
+end
+
+function _add_expression_tree_variable(
+    opcode::Vector{Cint},
+    data::Vector{Cdouble},
+    parent::Vector{Cint},
+    coeff::Float64,
+    var_index::Cint,
+    current_index::Cint,
+    parent_index::Cint,
+)
+    if coeff > 1.0
+        append!(opcode, GRB_OPCODE_MULTIPLY)
+        append!(data, -1.0)
+        append!(parent, parent_index)
+
+        _add_expression_tree_constant(
+            opcode,
+            data,
+            parent,
+            coeff,
+            current_index,
+        )
+
+        append!(opcode, GRB_OPCODE_VARIABLE)
+        append!(data, var_index)
+        append!(parent, current_index)
+        current_index += Cint(2)
+        return current_index
+    else
+        append!(data, var_index)
+        append!(opcode, GRB_OPCODE_VARIABLE)
+        append!(parent, parent_index)
+        current_index += Cint(1)
+        return current_index
+    end
+end
+
+# Check if a nonlinear is actually just a constant
+function _check_nonlinear_constant(expr)
+    return (
+        typeof(expr) == MOI.ScalarNonlinearFunction &&
+        expr.head in [:+, :-] &&
+        length(expr.args) == 1 &&
+        typeof(expr.args[1]) == Float64
+    )
+end
+
+# Check if a nonlinear is actually just an affine expression
+# Cases:
+#   1. constant * var
+#   2. +/- var
+#   3. +/- constant * var
+function _check_nonlinear_singlevar(expr)
+    return typeof(expr) == MOI.ScalarNonlinearFunction && (
+        ( # Case 1.
+            expr.head == :* &&
+            length(expr.args) == 2 &&
+            typeof(expr.args[2]) == MOI.VariableIndex
```

Comment:
I don't think this actually checks case 1? what if it is variable * variable?

Reply:
Won't variable * variable arrive in _process_nonlinear as a ScalarQuadraticFunction?

Reply:
Not always:
```julia
julia> model = Model();

julia> @variable(model, x)
x

julia> @objective(model, Min, @force_nonlinear(x * x))
x * x

julia> model
A JuMP Model
├ solver: none
├ objective_sense: MIN_SENSE
│ └ objective_function_type: NonlinearExpr
├ num_variables: 1
├ num_constraints: 0
└ Names registered in the model
  └ :x

julia> objective_function(model)
x * x

julia> typeof(objective_function(model))
NonlinearExpr (alias for GenericNonlinearExpr{GenericVariableRef{Float64}})

julia> objective_function(model).head
:*

julia> objective_function(model).args
2-element Vector{Any}:
 x
 x
```

Reply:
Hmm. Given that all the MOI tests are passing (when I run this locally), is it possible some important ones are still being skipped? Or is the coverage just not that good for NLP?

Reply:
> Or is the coverage just not that good for NLP?

The coverage just isn't as good. Until the introduction of `ScalarNonlinearFunction`, we didn't have a good way of writing `MOI.Test.` for nonlinear. 

There are a bunch of other tests in https://github.com/jump-dev/MINLPTests.jl. See KNITRO for an example of how to call: https://github.com/jump-dev/KNITRO.jl/blob/master/test/MINLPTests/run_minlptests.jl

Reply:
> https://github.com/jump-dev/MINLPTests.jl

Neat, thanks!

Reply:
>  Won't variable * variable arrive in _process_nonlinear as a ScalarQuadraticFunction?

I also thought this would be the case. Thanks for double checking this!

Reply:
> Hmm. Given that all the MOI tests are passing (when I run this locally)

I think the `test_nonlinear_scalarquadraticfunction` is failing because of this.

Reply:
Ah, right you are. I added that test thinking it would catch the missing dispatch of ScalarQuadraticFunction in the _process_nonlinear loop. But I see it actually catches this case. That explains why it didn't give the error message I expected!

Reply:
@odow from JuMP's perspective, is there actually any need for this special handling?

i.e. if presented with an explicit nonlinear function `:* {1.0, x}`, is it necessary to discard the parent node? If there's a ScalarAffineFunction or ScalarQuadraticFunction in the tree, it would make sense to come up with a simplified representation where possible, but for explicit nonlinear I'd guess it's preferable to represent it more or less faithfully to the solver.

Reply:
I don't know that I understand your question. JuMP doesn't have any perspective. It gives you what the user wrote. And it's up to you to do with that as you please.

I haven't looked in enough detail at why you're doing these calls.

But yes, I generally agree what it's preferable to pass the solver more or less what the user wrote.

Reply:
For example we can see both of these appear in the processing of a nonlinear function:
1. A ScalarAffineFunction with one term `1.0 * x`
2. A ScalarNonlinearFunction with `head=:*` and two children `Float{1.0}` and `VariableIndex`

I was wondering whether JuMP expects a different representation of those two objects in the underlying solver. Sounds like there are no requirements there.

Reply:
JuMP doesn't care